File: scripts/rust/tests/review_integration_test_structure.py
#!/usr/bin/env python3
"""
Review: Integration test structure.

RustRules.md Lines 292-298: "Integration tests must have test functions at the 
root level of the file. NEVER use #[cfg(test)] modules in integration test files
- this prevents test discovery."

Checks all files in tests/ directory for #[cfg(test)] module usage.
"""
# Git commit: 584a672b6a34782766863c5f76a461d3297a741a
# Date: 2025-10-17 05:17:36 -0700


import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent.parent / "lib"))
from review_utils import ReviewContext, create_review_parser


def check_file(file_path: Path, context: ReviewContext) -> list:
    """Check a single test file for #[cfg(test)] usage."""
    violations = []
    
    with open(file_path, 'r', encoding='utf-8') as f:
        in_multiline_comment = False
        for line_num, line in enumerate(f, start=1):
            # Handle multi-line comments
            if '/*' in line:
                in_multiline_comment = True
            if '*/' in line:
                in_multiline_comment = False
                continue
            if in_multiline_comment:
                continue
            
            stripped = line.strip()
            if stripped.startswith('//'):
                continue
            
            # Check for #[cfg(test)]
            if '#[cfg(test)]' in line:
                rel_path = context.relative_path(file_path)
                violations.append(f"  {rel_path}:{line_num}\n    {stripped}")
    
    return violations


def main():
    parser = create_review_parser(__doc__)
    args = parser.parse_args()
    context = ReviewContext(args)
    
    tests_dir = context.repo_root / "tests"
    if not tests_dir.exists():
        print("✓ No tests/ directory found")
        return 0
    
    if context.dry_run:
        files = context.find_files([tests_dir])
        print(f"Would check {len(files)} file(s) for #[cfg(test)] usage")
        return 0
    
    all_violations = []
    files = context.find_files([tests_dir])
    
    for file_path in files:
        violations = check_file(file_path, context)
        all_violations.extend(violations)
    
    if not all_violations:
        print("✓ No #[cfg(test)] modules in integration tests")
        return 0
    
    print(f"✗ Found #[cfg(test)] in integration tests (RustRules.md Lines 292-298):\n")
    for violation in all_violations:
        print(violation)
    print(f"\nTotal violations: {len(all_violations)}")
    print("\nFix: Remove #[cfg(test)] modules from integration tests.")
    print("Integration tests should have #[test] functions at root level.")
    return 1


if __name__ == "__main__":
    sys.exit(main())

File: src/bin/review_integration_test_structure.rs
// Copyright (C) Brian G. Milnes 2025

//! Review: Integration test structure (MANDATORY)
//! 
//! Replaces: scripts/rust/tests/review_integration_test_structure.py
//! Rule: RustRules.md Lines 292-298
//! "Integration tests must have test functions at the root level of the file.
//! NEVER use #[cfg(test)] modules in integration test files - this prevents test discovery."
//! Git commit: 584a672b6a34782766863c5f76a461d3297a741a
//! Binary: rusticate-review-integration-test-structure
use std::fs;
use std::path::{Path, PathBuf};
use std::time::Instant;

use anyhow::Result;
use rusticate::{StandardArgs, format_number, find_rust_files};


macro_rules! log {
    ($($arg:tt)*) => {{
        use std::io::Write;
        let msg = format!($($arg)*);
        println!("{}", msg);
        if let Ok(mut file) = std::fs::OpenOptions::new()
            .create(true)
            .append(true)
            .open("analyses/review_integration_test_structure.log")
        {
            let _ = writeln!(file, "{}", msg);
        }
    }};
}
#[derive(Debug)]
struct Violation {
    file: PathBuf,
    line_num: usize,
    line_content: String,
}

fn check_file(file_path: &Path) -> Result<Vec<Violation>> {
    let content = fs::read_to_string(file_path)?;
    let mut violations = Vec::new();
    
    let mut in_multiline_comment = false;
    
    for (line_idx, line) in content.lines().enumerate() {
        let line_num = line_idx + 1;
        
        // Handle multi-line comments
        if line.contains("/*") {
            in_multiline_comment = true;
        }
        if line.contains("*/") {
            in_multiline_comment = false;
            continue;
        }
        if in_multiline_comment {
            continue;
        }
        
        let trimmed = line.trim();
        
        // Skip single-line comments
        if trimmed.starts_with("//") {
            continue;
        }
        
        // Check for #[cfg(test)]
        if line.contains("#[cfg(test)]") {
            violations.push(Violation {
                file: file_path.to_path_buf(),
                line_num,
                line_content: trimmed.to_string(),
            });
        }
    }
    
    Ok(violations)
}

fn main() -> Result<()> {
    let start = Instant::now();
    let args = StandardArgs::parse()?;
    let base_dir = args.base_dir();
    
    // Print compilation directory for Emacs compile-mode
    log!("Entering directory '{}'", base_dir.display());
    log!("");
    
    let search_dirs = args.get_search_dirs();
    
    // This rule only applies to tests/ directory
    let tests_dirs: Vec<_> = search_dirs.iter()
        .filter(|p| p.is_dir() && (p.ends_with("tests") || p.components().any(|c| c.as_os_str() == "tests")))
        .cloned()
        .collect();
    
    if tests_dirs.is_empty() {
        log!("✓ No tests/ directory found");
        let elapsed = start.elapsed().as_millis();
        log!("Completed in {}ms", elapsed);
        return Ok(());
    }
    
    let files = find_rust_files(&tests_dirs);
    let mut all_violations = Vec::new();
    
    for file in &files {
        match check_file(file) {
            Ok(violations) => all_violations.extend(violations),
            Err(e) => {
                eprintln!("Warning: Failed to read {}: {}", file.display(), e);
            }
        }
    }
    
    // Report violations
    if all_violations.is_empty() {
        log!("✓ No #[cfg(test)] modules in integration tests");
    } else {
        log!("✗ Found #[cfg(test)] in integration tests (RustRules.md Lines 292-298):");
        log!("");
        
        for v in &all_violations {
            if let Ok(rel_path) = v.file.strip_prefix(&base_dir) {
                log!("{}:{}: #[cfg(test)] in integration test", 
                         rel_path.display(), v.line_num);
                log!("  {}", v.line_content);
            }
        }
        
        log!("");
        log!("Fix: Remove #[cfg(test)] modules from integration tests.");
        log!("Integration tests should have #[test] functions at root level.");
    }
    
    // Summary
    log!("");
    log!("Summary: {} files checked, {} violations", 
             format_number(files.len()), format_number(all_violations.len()));
    
    let elapsed = start.elapsed().as_millis();
    log!("Completed in {}ms", elapsed);
    
    if all_violations.is_empty() {
        Ok(())
    } else {
        std::process::exit(1);
    }
}

